File: ios/Classes/SuperwallPlugin.swift
import Flutter
import UIKit
import SuperwallKit

public class SuperwallPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "SWK_SuperwallPlugin", binaryMessenger: registrar.messenger())
    let instance = SuperwallPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
      case "getDelegate":
        // Implement logic to get the delegate
        let delegate = Superwall.shared.delegate
        result(delegate)

      case "setDelegate":
        // Implement logic to set the delegate
        if let args = call.arguments as? [String: Any], let delegate = args["delegate"] as? SuperwallDelegate {
          Superwall.shared.delegate = delegate
        }
        result(nil)

      case "getLogLevel":
        // Implement logic to get log level
        result(Superwall.shared.logLevel.rawValue)

      case "setLogLevel":
        // Implement logic to set log level
        if let args = call.arguments as? [String: Any], let level = args["logLevel"] as? Int, let logLevel = LogLevel(rawValue: level) {
          Superwall.shared.logLevel = logLevel
        }
        result(nil)

      case "getUserAttributes":
        // Implement logic to get user attributes
        result(Superwall.shared.userAttributes)

      case "getUserId":
        // Implement logic to get the current user's id
        result(Superwall.shared.userId)

      case "getIsLoggedIn":
        // Implement logic to check if the user is logged in to Superwall
        result(Superwall.shared.isLoggedIn)

      case "getPresentedViewController":
        // Implement logic to get the presented paywall view controller
        // TODO: Since UIViewController cannot be returned directly to Dart, handle appropriately
        if let viewController = Superwall.shared.presentedViewController {
          result(viewController.description)
        } else {
          result(nil)
        }

      case "getLatestPaywallInfo":
        // Implement logic to get the latest PaywallInfo object
        if let paywallInfo = Superwall.shared.latestPaywallInfo {
          // TODO: Convert PaywallInfo to a format suitable for passing over method channels
          result(nil)
        } else {
          result(nil)
        }

      case "getSubscriptionStatus":
        // Implement logic to get the subscription status of the user
        result(Superwall.shared.subscriptionStatus.rawValue)

      case "setSubscriptionStatus":
        // Implement logic to set the subscription status of the user
        if let args = call.arguments as? [String: Any], let statusRawValue = args["status"] as? Int, let status = SubscriptionStatus(rawValue: statusRawValue) {
          Superwall.shared.subscriptionStatus = status
        }
        result(nil)

      case "getIsConfigured":
        // Implement logic to check if Superwall has finished configuring
        result(Superwall.shared.isConfigured)

      case "setIsConfigured":
        // Implement logic to set the configured state of Superwall
        if let args = call.arguments as? [String: Any], let configured = args["configured"] as? Bool {
          Superwall.shared.isConfigured = configured
        }
        result(nil)

      case "getIsPaywallPresented":
        // Implement logic to check if a paywall is currently being presented
        result(Superwall.shared.isPaywallPresented)

      case "preloadAllPaywalls":
        // Implement logic to preload all paywalls
        Superwall.shared.preloadAllPaywalls()
        result(nil)

      case "preloadPaywallsForEvents":
        // Implement logic to preload paywalls for specific event names
        if let args = call.arguments as? [String: Any], let eventNames = args["eventNames"] as? [String] {
          Superwall.shared.preloadPaywalls(forEvents: Set(eventNames))
        }
        result(nil)

      case "handleDeepLink":
        // Implement logic to handle deep links for paywall previews
        if let args = call.arguments as? [String: Any], let urlString = args["url"] as? String, let url = URL(string: urlString) {
          let handled = Superwall.shared.handleDeepLink(url)
          result(handled)
        } else {
          result(FlutterError(code: "INVALID_URL", message: "Invalid URL provided", details: nil))
        }

      case "togglePaywallSpinner":
        // Implement logic to toggle the paywall loading spinner
        if let args = call.arguments as? [String: Any], let isHidden = args["isHidden"] as? Bool {
          Superwall.shared.togglePaywallSpinner(isHidden: isHidden)
        }
        result(nil)

      case "reset":
        // Implement logic to reset the user ID, on-device paywall assignments, and stored data
        Superwall.shared.reset()
        result(nil)

      case "configure":
        // Implement logic to configure the Superwall instance
        guard let apiKey: String = call.argument(for: "apiKey_iOS") else {
          result(nil)
          return
        }

        let purchaseController: PurchaseController? = call.argument(for: "purchaseController")
        let options: SuperwallOptions? = call.argument(for: "options")

        Superwall.configure(apiKey: apiKey, purchaseController: purchaseController, options: options)

        // Returning nil instead of the result from configure because we want to use the Dart
        // instance of Superwall, not a native variant
        result(nil)

      default:
        result(FlutterMethodNotImplemented)
    }
  }
}

File: android/src/main/kotlin/com/superwall/superwallkit_flutter/SuperwallPlugin.kt
package com.superwall.superwallkit_flutter

import android.app.Activity
import com.superwall.sdk.Superwall
import com.superwall.sdk.config.options.SuperwallOptions
import com.superwall.sdk.delegate.SuperwallDelegate
import com.superwall.sdk.delegate.subscription_controller.PurchaseController
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import android.net.Uri
import com.superwall.sdk.delegate.SubscriptionStatus
import com.superwall.sdk.misc.ActivityProvider

class SuperwallPlugin: FlutterPlugin, MethodChannel.MethodCallHandler, ActivityProvider {
    private lateinit var channel: MethodChannel
    private lateinit var flutterPluginBinding: FlutterPlugin.FlutterPluginBinding

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        this.flutterPluginBinding = flutterPluginBinding
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "SWK_SuperwallBridge")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getDelegate" -> {
                val delegate = Superwall.instance.delegate
                result.success(delegate)
            }
            "setDelegate" -> {
                val delegate = call.argumentForKey<SuperwallDelegate?>("delegate")
                Superwall.instance.delegate = delegate
                result.success(null)
            }
            "getLogLevel" -> {
                // TODO
//                val logLevel = Superwall.instance.logLevel
//                result.success(logLevel.ordinal)
                result.notImplemented()
            }
            "setLogLevel" -> {
                // TODO
//                val levelOrdinal = call.argumentForKey<Int>("logLevel")
//                levelOrdinal?.let {
//                    val logLevel = LogLevel.values()[it]
//                    Superwall.instance.logLevel = logLevel
//                }
//                result.success(null)
                result.notImplemented()
            }
            "getUserAttributes" -> {
                // TODO: Fix IdentityManager class in Android
//                val attributes = Superwall.instance.getUserAttributes()
//                result.success(attributes)
                result.notImplemented()
            }
            "getUserId" -> {
                // TODO: Add to Android
//                val userId = Superwall.instance.userId
//                result.success(userId)
                result.notImplemented()
            }
            "getIsLoggedIn" -> {
                // TODO: Add to Android
//                val isLoggedIn = Superwall.instance.isLoggedIn
//                result.success(isLoggedIn)
                result.notImplemented()
            }
            "getPresentedViewController" -> {
                // TODO: Since UIViewController cannot be returned directly to Dart, handle appropriately
//                val viewController = Superwall.instance.presentedViewController
//                result.success(viewController?.description)
                result.notImplemented()
            }
            "getLatestPaywallInfo" -> {
                // TODO: Convert PaywallInfo to a format suitable for passing over method channels
//                val paywallInfo = Superwall.instance.latestPaywallInfo
//                result.success(paywallInfo?.toMap())
                result.notImplemented()
            }
            "getSubscriptionStatus" -> {
                val statusStateFlow = Superwall.instance.subscriptionStatus
                val status = statusStateFlow.value
                result.success(status.rawValue)
            }
            "setSubscriptionStatus" -> {
                val statusRawValue = call.argumentForKey<Int>("status")
                statusRawValue?.let {
                    val status = SubscriptionStatus.fromRawValue(statusRawValue)
                    Superwall.instance.setSubscriptionStatus(status)
                }
                result.success(null)
            }
            "getIsConfigured" -> {
                // TODO: Add to Android
//                val isConfigured = Superwall.instance.isConfigured
//                result.success(isConfigured)
                result.notImplemented()
            }
            "setIsConfigured" -> {
                // TODO: Add to Android
//                val configured = call.argumentForKey<Boolean>("configured")
//                configured?.let {
//                    Superwall.instance.isConfigured = it
//                }
//                result.success(null)
                result.notImplemented()
            }
            "getIsPaywallPresented" -> {
                val isPaywallPresented = Superwall.instance.isPaywallPresented
                result.success(isPaywallPresented)
            }
            "preloadAllPaywalls" -> {
                // TODO: Add to Android
//                Superwall.instance.preloadAllPaywalls()
//                result.success(null)
                result.notImplemented()
            }
            "preloadPaywallsForEvents" -> {
                // TODO: Add to Android
//                val eventNames = call.argumentForKey<List<String>>("eventNames")?.toSet()
//                eventNames?.let {
//                    Superwall.instance.preloadPaywalls(forEvents: it)
//                }
//                result.success(null)
                result.notImplemented()
            }
            "handleDeepLink" -> {
                val urlString = call.argument<String>("url")
                urlString?.let {
                    try {
                        val uri = Uri.parse(it)
                        val handled = Superwall.instance.handleDeepLink(uri)
                        result.success(handled)
                    } catch (e: Exception) {
                        // Handle any other exceptions during parsing
                        result.error("URI_PARSE_ERROR", "Error parsing URL: ${e.message}", null)
                    }
                } ?: run {
                    // urlString is null
                    result.error("INVALID_URL", "URL is null", null)
                }
            }
            "togglePaywallSpinner" -> {
                val isHidden = call.argumentForKey<Boolean>("isHidden")
                isHidden?.let {
                    Superwall.instance.togglePaywallSpinner(isHidden = it)
                }
                result.success(null)
            }
            "reset" -> {
                Superwall.instance.reset()
                result.success(null)
            }
            "configure" -> {
                val apiKey = call.argumentForKey<String>("apiKey_android")
                apiKey?.let {
                    // TODO: need primitives for purchase controller and options
                    val purchaseController = call.argumentForKey<PurchaseController?>("purchaseController")
                    val options = call.argumentForKey<SuperwallOptions?>("options")
                    Superwall.configure(flutterPluginBinding.applicationContext, apiKey = it, purchaseController = purchaseController, options = options, activityProvider = this)
                }
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun getCurrentActivity(): Activity? {
        return SuperwallkitFlutterPlugin.currentActivity
    }
}